### PR TITLE
Mol 12297

### PR DESCRIPTION
**Pull request by CWH-UF**
**Changes**
Description of changes...

- use textmetrics to get filename element width 
- updated to use while loop and context to generate truncate filename

**Additional information**

-   [MOL-12297](https://jira.ship.gov.sg/browse/MOL-12297)

### CODE REVIEW

**Comment by weili-govtech** [comment]
instead of introducing `fileNameElementWidth` and a hidden field, may I propose to:
- call `setFileNameToWidth` when image is loaded, I think this can be done via resize observer
- apply `word-break: break-all;` to TextBody to forcefully constrain its width to its parent container

**Comment by ryhuz** [comment]
your commit message does not match your change anymore

**Review by ryhuz — CHANGES_REQUESTED** [review]
your commit message does not match your change anymore

File: src/utils/file-helper.ts [hunk 1]
```diff
@@ -4,18 +4,64 @@ export namespace FileHelper {
...
-		// TODO: how to better determine char width?
-		const threshold = Math.floor(widthOfElement / 8);
-		const ellipsis = "...";
-		const length = fileName.length + ellipsis.length;
-		if (length > threshold) {
-			const thresholdIndex = threshold / 2;
-			return fileName.substring(0, thresholdIndex) + ellipsis + fileName.substring(length - thresholdIndex);
+	export const truncateFileName = (
+		fileName: string,
+		ref?: React.MutableRefObject<HTMLDivElement | HTMLParagraphElement>
```
Comment (weili-govtech) [inline src/utils/file-helper.ts:9]:
```suggestion
		ref?: React.MutableRefObject<HTMLDivElement | HTMLParagraphElement> | undefined
```

File: src/utils/file-helper.ts [hunk 2]
```diff
@@ -4,18 +4,64 @@ export namespace FileHelper {
...
+	) => {
+		let truncateFileName = fileName;
+		let widthOfElement = 0;
+		let context;
```
Comment (weili-govtech) [inline src/utils/file-helper.ts:13]:
```suggestion
		let context: CanvasRenderingContext2D;
```

File: src/utils/file-helper.ts [hunk 3]
```diff
@@ -4,18 +4,64 @@ export namespace FileHelper {
...
-		const ellipsis = "...";
-		const length = fileName.length + ellipsis.length;
-		if (length > threshold) {
-			const thresholdIndex = threshold / 2;
-			return fileName.substring(0, thresholdIndex) + ellipsis + fileName.substring(length - thresholdIndex);
+	export const truncateFileName = (
+		fileName: string,
+		ref?: React.MutableRefObject<HTMLDivElement | HTMLParagraphElement>
+	) => {
+		let truncateFileName = fileName;
```
Comment (weili-govtech) [inline src/utils/file-helper.ts:11]: avoid naming the variable the same as the function
```suggestion
		let truncatedFileName = fileName;
```

File: src/utils/file-helper.ts (hunk 1 above)
Comment (weili-govtech) [inline src/utils/file-helper.ts:9]:
```suggestion
		ref?: React.MutableRefObject<HTMLDivElement | HTMLParagraphElement> | undefined
```

**Review by weili-govtech — CHANGES_REQUESTED** [review]

File: src/utils/file-helper.ts (hunk 3 above)
Comment (CWH-UF) [reply src/utils/file-helper.ts]: updated

File: src/utils/file-helper.ts (hunk 1 above)
Comment (CWH-UF) [reply src/utils/file-helper.ts]: updated

Comment (CWH-UF) [reply src/utils/file-helper.ts]: updated

**Review by weili-govtech — DISMISSED** [review]

**Comment by CWH-UF** [comment]
updated while loop variable `i & j` to `startIndex & endIndex`

**Review by weili-govtech — APPROVED** [review]

**Review by ryhuz — APPROVED** [review]

**Pull request merged**